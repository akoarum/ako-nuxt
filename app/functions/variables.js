export const VALIDATES = {
  EMPTY: {
    message: '入力してください。'
  },
  NUMBER: {
    regex: /^[0-9]+?$/,
    message: '数字で入力してください。'
  },
  EMAIL: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスの形式になっていません。'
  },
  TEL: {
    regex: /^[0-9]{2,4}-?[0-9]{2,4}-?[0-9]{4}$/,
    message: '電話番号の形式になっていません。'
  },
  URL: {
    regex: /^http(s):\/\/.+?\.+?$/,
    message: 'URLの形式になっていません。'
  }
}
