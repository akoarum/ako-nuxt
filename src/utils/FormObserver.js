export default class FormValidator {
  constructor(names = []) {
    this._items = this._createItems(names)
  }

  /**
   * @return {Boolean}
   */
  get btnDisabled() {
    for (const name in this._items) {
      if (this._items[name].error || !this._items[name].dirty) return true
    }
    return false
  }

  /**
   * itemオブジェクトの生成
   * @param {Array} names name配列
   */
  _createItems(names) {
    const items = {}

    names.forEach(item => {
      items[item] = {
        error: false,
        dirty: false
      }
    })

    return items
  }

  /**
   * itemの更新
   * @param {String}  name  name値
   * @param {String}  prop  更新するプロパティ名
   * @param {Boolean} value 値
   */
  updateItem(name, prop, value) {
    this._items[name][prop] = value
  }
}
