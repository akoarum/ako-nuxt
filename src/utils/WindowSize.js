export default class WindowSize {
  constructor(delay = 80) {
    this._width = 0
    this._height = 0
    this._delay = delay
    this._cullTime = null
    this._initialized = false
    this._handler = this._handleResize.bind(this)
  }

  get width() {
    return this._width
  }

  get height() {
    return this._height
  }

  get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  initialize() {
    if (this._initialized) return this
    this._updateState()
    window.addEventListener('resize', this._handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  _updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  _handleResize() {
    clearTimeout(this._cullTime)
    this._cullTime = setTimeout(() => {
      this._updateState()
    }, this._delay)
  }

  /**
   * 解放する
   */
  destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this._handler)
    this._initialized = false
    return this
  }
}
