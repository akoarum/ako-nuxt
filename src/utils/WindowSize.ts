export default class WindowSize {
  private _width: number = 0
  private _height: number = 0
  private delay: number = 80
  private cullTime = null
  private _initialized: boolean = false
  private handler = this.handleResize.bind(this)

  constructor(delay: number = 80) {
    this.delay = delay
  }

  public get width() {
    return this._width
  }

  public get height() {
    return this._height
  }

  public get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  public initialize() {
    if (this._initialized) return this
    this.updateState()
    window.addEventListener('resize', this.handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  public updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  private handleResize() {
    clearTimeout(this.cullTime)
    this.cullTime = setTimeout(() => {
      this.updateState()
    }, this.delay)
  }

  /**
   * 解放する
   */
  public destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this.handler)
    this._initialized = false
    return this
  }
}
