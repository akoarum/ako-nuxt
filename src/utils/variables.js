export const VALIDATES = {
  required: '入力してください。',
  selected: '選択してください。',
  file: 'ファイルをアップロードしてください。',
  tel: {
    regex: /^\d{2,4}-?\d{4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
