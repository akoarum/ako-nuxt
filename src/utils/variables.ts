import { ValidatePattern } from '~/types/models'

export const VALIDATE_PATTERNS: { [key: string]: ValidatePattern } = {
  required: {
    message: '入力してください。'
  },
  selected: {
    message: '選択してください。'
  },
  tel: {
    regex: /^\d{2,5}-?\d{1,4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
