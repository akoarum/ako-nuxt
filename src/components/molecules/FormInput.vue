<template>
  <div class="formInput">
    <VInput
      v-model="model"
      v-bind="$attrs"
      :type="type"
      :required="required"
      :aria-invalid="String(isError)"
      :aria-describedby="isError ? `error-${name}` : false"
      class="formInput__input"
      @focus="$emit('focus')"
      @blur="handleBlur"
    />
    <template v-if="isError">
      <VTexts
        :id="`error-${name}`"
        class="formInput__error"
        role="alert"
        aria-live="assertive"
        small
        caution
      >
        <span>{{ error }}</span>
      </VTexts>
    </template>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { VALIDATE_PATTERNS } from '~/utils/variables'
import { VInput, VTexts } from '~/components/atoms'

type Data = {
  error: string
  isDirty: boolean
}

export default Vue.extend({
  components: { VInput, VTexts },
  inheritAttrs: false,
  props: {
    name: { type: String, required: true },
    value: { type: String, required: true },
    type: { type: String, default: 'text' },
    required: { type: Boolean },
    validateType: { type: String }
  },
  data(): Data {
    return {
      error: '',
      isDirty: false
    }
  },
  computed: {
    model: {
      get(): string {
        return this.value
      },
      set(value: string) {
        this.$emit('input', value)
        this.error = this.validate(value)
      }
    },
    isError(): boolean {
      return !!this.error
    }
  },
  watch: {
    isDirty() {
      this.$emit('dirty')
      this.error = this.validate(this.value)
    },
    isError(newValue: boolean) {
      this.$emit('error', newValue)
    }
  },
  created() {
    if (this.value || !this.required) {
      this.updateDirty()
    }
  },
  methods: {
    handleBlur() {
      this.$emit('blur')

      if (!this.isDirty) {
        this.updateDirty()
      }
    },
    updateDirty() {
      if (this.isDirty) return
      if (this.value) {
        this.isDirty = true
      }
    },
    validate(value: string): string {
      if (!this.isDirty) return ''
      if (!this.required && !value) return ''
      if (this.required && !value) return VALIDATE_PATTERNS.required.message

      for (const type in VALIDATE_PATTERNS) {
        if (this.type === type) {
          if (value && !VALIDATE_PATTERNS[type].regex!.test(value)) {
            return VALIDATE_PATTERNS[type].message
          }
        }
      }

      return ''
    }
  }
})
</script>

<style lang="scss" scoped>
.formInput__error {
  @include media-min {
    margin-top: 6px;
  }
  @include media-max {
    margin-top: 8px;
  }
}
</style>
