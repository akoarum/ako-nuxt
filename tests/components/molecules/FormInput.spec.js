import { shallowMount } from '@vue/test-utils'
import FormInput from '~/components/molecules/FormInput'
import { VALIDATES } from '~/utils/variables'

describe('FormInput', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(FormInput, {
      propsData: {
        name: 'name',
        type: 'text',
        value: '',
        required: true
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('validate', () => {
      it('引数valueが空でrequiredでない場合は空文字列を返す', () => {
        wrapper.setProps({ required: false })
        expect(wrapper.vm.validate('')).toBe('')
      })

      it('引数valueが空でrequired、!isDirtyの場合は空文字列を返す', () => {
        expect(wrapper.vm.validate('')).toBe('')
      })

      it('引数valueが空でrequired、isDirtyの場合はrequiredを返す', () => {
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('')).toBe(VALIDATES.required)
      })

      it('type: email、引数valueがEMAILになっていない場合はemail.messageを返す', () => {
        wrapper.setProps({ type: 'email' })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('test')).toBe(VALIDATES.email.message)
      })

      it('type: tel、引数valueがTELになっていない場合はtel.messageを返す', () => {
        wrapper.setProps({ type: 'tel' })
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('test')).toBe(VALIDATES.tel.message)
      })

      it('上記以外であれば空文字列を返す', () => {
        wrapper.setData({ isDirty: true })
        expect(wrapper.vm.validate('test')).toBe('')
      })
    })

    describe('updateDirty', () => {
      it('valueが空でなくisDirtyでなければisDirtyをtrueにする', () => {
        const mockValidate = jest.fn()
        wrapper.setProps({ value: 'test' })
        wrapper.setMethods({ validate: mockValidate })
        wrapper.vm.updateDirty()
        expect(wrapper.vm.isDirty).toBe(true)
        expect(mockValidate).toHaveBeenCalled()
      })
    })
  })
})
