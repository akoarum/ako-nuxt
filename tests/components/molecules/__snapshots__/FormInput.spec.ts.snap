// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput initialized snapshot 1`] = `
<div
  class="formInput"
>
  <vinput-stub
    aria-invalid="false"
    class="formInput__input"
    type="text"
    value=""
  />
   
  <!---->
</div>
`;
